src/NuGet.Clients/NuGet.SolutionRestoreManager/ProjectRestoreInfo3Adapter.cs

```diff
@@ -0,0 +1,210 @@
+            {
+                var property = properties.Item(i) ?? throw new Exception($"Property index {i} returned null");
+                result.Add(property.Name, property.Value);
+            }
+
+            return result;
+        }
+
+        private static IReadOnlyDictionary<string, string?>? ToPropertiesDictionary(IVsReferenceProperties? properties)
+        {
+            if (properties is null) { return null; }
+
+            var result = new Dictionary<string, string?>(properties.Count, StringComparer.OrdinalIgnoreCase);
+
+            for (var i = 0; i < properties.Count; i++)
+            {
+                var property = properties.Item(i) ?? throw new Exception($"Property index {i} returned null");
+                result.Add(property.Name, property.Value);
+            }
+
+            return result;
+        }
+
+        private record TargetFramework4Adapter : IVsTargetFrameworkInfo4
+        {
+            public TargetFramework4Adapter(IVsTargetFrameworkInfo targetFrameworkInfo)
+            {
+                if (targetFrameworkInfo is null) { throw new ArgumentNullException(nameof(targetFrameworkInfo)); }
+
+                TargetFrameworkMoniker = targetFrameworkInfo.TargetFrameworkMoniker;
+                Items = ToItemsDictionary(targetFrameworkInfo);
+                // null needs to be handled in VsSolutionRestoreService.ToDependencySpec and reported back to caller
+                Properties = targetFrameworkInfo.Properties is not null ? ToPropertiesDictionary(targetFrameworkInfo.Properties) : null!;
+            }
+
+            private IReadOnlyDictionary<string, IReadOnlyList<IVsReferenceItem2>> ToItemsDictionary(IVsTargetFrameworkInfo targetFrameworkInfo)
+            {
+                var result = new Dictionary<string, IReadOnlyList<IVsReferenceItem2>>(StringComparer.OrdinalIgnoreCase);
+
+                if (targetFrameworkInfo.ProjectReferences?.Count > 0)
+                {
+                    IReadOnlyList<IVsReferenceItem2> projectReferences = ToReferenceItem2ReadOnlyList(targetFrameworkInfo.ProjectReferences);
+                    result[ProjectItems.ProjectReference] = projectReferences;
+                }
+
+                if (targetFrameworkInfo.PackageReferences?.Count > 0)
+                {
+                    IReadOnlyList<IVsReferenceItem2> packageReferences = ToReferenceItem2ReadOnlyList(targetFrameworkInfo.PackageReferences);
+                    result[ProjectItems.PackageReference] = packageReferences;
+                }
+
+                if (targetFrameworkInfo is IVsTargetFrameworkInfo2 targetFrameworkInfo2)
+                {
+                    if (targetFrameworkInfo2.PackageDownloads?.Count > 0)
+                    {
+                        IReadOnlyList<IVsReferenceItem2> packageDownloads = ToReferenceItem2ReadOnlyList(targetFrameworkInfo2.PackageDownloads);
+                        result["PackageDownload"] = packageDownloads;
```

Comment:
```suggestion
                        result[ProjectItems.PackageDownload] = packageDownloads;
```

Replaced hardcoded strings with constants. I see a few other hardcoded strings in the changeset; maybe we can replace them with constants.